File: agents/go/config_test.go
package openlog

import (
	"log/slog"
	"strings"
	"testing"
	"time"
)

func env(m map[string]string) lookupFunc {
	return func(k string) (string, bool) { v, ok := m[k]; return v, ok }
}

func TestConfigPrecedence(t *testing.T) {
	full := map[string]string{
		"OTEL_SERVICE_NAME":           "otel-svc",
		"OPENLOG_SERVICE_NAME":        "openlog-svc",
		"OTEL_EXPORTER_OTLP_ENDPOINT": "http://otel:4318",
		"OPENLOG_ENDPOINT":            "http://openlog:4318",
		"OTEL_TRACES_SAMPLER":         "parentbased_traceidratio",
		"OTEL_TRACES_SAMPLER_ARG":     "0.5",
		"OPENLOG_SAMPLING_RATIO":      "0.25",
		"OTEL_LOG_LEVEL":              "error",
		"OPENLOG_LOG_LEVEL":           "debug",
		"OTEL_RESOURCE_ATTRIBUTES":    "team=otel,region=eu",
		"OPENLOG_RESOURCE_ATTRIBUTES": "team=openlog",
		"OTEL_EXPORTER_OTLP_HEADERS":  "x-a=1,openlog-license-key=from-otel-headers",
		"OPENLOG_LICENSE_KEY":         "lk",
		"OPENLOG_ENVIRONMENT":         "prod",
	}
	cases := []struct {
		name  string
		env   map[string]string
		opts  []Option
		check func(t *testing.T, c *Config)
	}{
		{"option beats env", full, []Option{WithServiceName("opt-svc"), WithEndpoint("http://opt:4318"), WithSamplingRatio(0.1)},
			func(t *testing.T, c *Config) {
				eq(t, c.ServiceName, "opt-svc")
				eq(t, c.Endpoint, "http://opt:4318")
				eq(t, c.SamplingRatio, 0.1)
			}},
		{"OPENLOG beats OTEL", full, nil, func(t *testing.T, c *Config) {
			eq(t, c.ServiceName, "openlog-svc")
			eq(t, c.Endpoint, "http://openlog:4318")
			eq(t, c.SamplingRatio, 0.25)
			eq(t, c.LogLevel, slog.LevelDebug)
			eq(t, c.ResourceAttributes["team"], "openlog")
			eq(t, c.ResourceAttributes["region"], "eu")
			eq(t, c.Headers["x-a"], "1")
			eq(t, c.Headers[LicenseKeyHeader], "lk")
			eq(t, c.Environment, "prod")
		}},
		{"OTEL only", map[string]string{"OTEL_SERVICE_NAME": "otel-svc", "OTEL_EXPORTER_OTLP_ENDPOINT": "http://otel:4318",
			"OTEL_TRACES_SAMPLER_ARG": "0.5", "OTEL_LOG_LEVEL": "error", "OTEL_METRIC_EXPORT_INTERVAL": "15000"},
			nil, func(t *testing.T, c *Config) {
				eq(t, c.ServiceName, "otel-svc")
				eq(t, c.Endpoint, "http://otel:4318")
				eq(t, c.SamplingRatio, 0.5)
				eq(t, c.LogLevel, slog.LevelError)
				eq(t, c.MetricInterval, 15*time.Second)
			}},
		{"sampler arg ignored for non-ratio sampler", map[string]string{"OTEL_TRACES_SAMPLER": "always_on", "OTEL_TRACES_SAMPLER_ARG": "0.5"},
			nil, func(t *testing.T, c *Config) { eq(t, c.SamplingRatio, 1.0) }},
		{"defaults", map[string]string{}, nil, func(t *testing.T, c *Config) {
			eq(t, c.Endpoint, "http://localhost:4318")
			eq(t, c.Protocol, ProtocolHTTP)
			eq(t, c.Compression, "gzip")
			eq(t, c.SamplingRatio, 1.0)
			eq(t, c.MetricInterval, 60*time.Second)
			eq(t, c.RuntimeMetrics, true)
			eq(t, c.Enabled, true)
			eq(t, c.LogLevel, slog.LevelWarn)
			if !strings.HasPrefix(c.ServiceName, "unknown_service:") {
				t.Errorf("service name %q", c.ServiceName)
			}
		}},
		{"grpc default endpoint", map[string]string{"OPENLOG_PROTOCOL": "grpc"}, nil, func(t *testing.T, c *Config) {
			eq(t, c.Endpoint, "http://localhost:4317")
			eq(t, c.Protocol, ProtocolGRPC)
		}},
		{"missing scheme means https", map[string]string{"OPENLOG_ENDPOINT": "ingest.example.com:4318/"}, nil, func(t *testing.T, c *Config) {
			eq(t, c.Endpoint, "https://ingest.example.com:4318")
		}},
		{"service.name from resource attributes", map[string]string{"OTEL_RESOURCE_ATTRIBUTES": "service.name=from-attrs"}, nil,
			func(t *testing.T, c *Config) { eq(t, c.ServiceName, "from-attrs") }},
		{"OTEL_SDK_DISABLED", map[string]string{"OTEL_SDK_DISABLED": "true"}, nil, func(t *testing.T, c *Config) { eq(t, c.Enabled, false) }},
		{"OPENLOG_ENABLED beats OTEL_SDK_DISABLED", map[string]string{"OTEL_SDK_DISABLED": "true", "OPENLOG_ENABLED": "true"}, nil,
			func(t *testing.T, c *Config) { eq(t, c.Enabled, true) }},
		{"option disables", map[string]string{"OPENLOG_ENABLED": "true"}, []Option{WithEnabled(false)},
			func(t *testing.T, c *Config) { eq(t, c.Enabled, false) }},
		{"ratio clamped", map[string]string{"OPENLOG_SAMPLING_RATIO": "3"}, nil, func(t *testing.T, c *Config) { eq(t, c.SamplingRatio, 1.0) }},
		{"durations and switches", map[string]string{"OPENLOG_METRIC_EXPORT_INTERVAL": "10s", "OPENLOG_SHUTDOWN_TIMEOUT": "2s",
			"OPENLOG_RUNTIME_METRICS": "false", "OPENLOG_COMPRESSION": "none", "OPENLOG_HOST_ID": "fixed"}, nil,
			func(t *testing.T, c *Config) {
				eq(t, c.MetricInterval, 10*time.Second)
				eq(t, c.ShutdownTimeout, 2*time.Second)
				eq(t, c.RuntimeMetrics, false)
				eq(t, c.Compression, "none")
				eq(t, c.HostID, "fixed")
			}},
		{"percent-encoded attribute values", map[string]string{"OPENLOG_RESOURCE_ATTRIBUTES": "a=b%20c, d = e "}, nil,
			func(t *testing.T, c *Config) {
				eq(t, c.ResourceAttributes["a"], "b c")
				eq(t, c.ResourceAttributes["d"], "e")
			}},
	}
	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			c, _, err := loadConfig(env(tc.env), tc.opts)
			if err != nil {
				t.Fatal(err)
			}
			tc.check(t, c)
		})
	}
}

func TestConfigErrorsAndWarnings(t *testing.T) {
	if _, _, err := loadConfig(env(map[string]string{"OPENLOG_PROTOCOL": "thrift"}), nil); err == nil {
		t.Error("unsupported protocol accepted")
	}
	if _, _, err := loadConfig(env(map[string]string{"OPENLOG_ENDPOINT": "ftp://x"}), nil); err == nil {
		t.Error("ftp endpoint accepted")
	}
	if _, _, err := loadConfig(env(map[string]string{"OPENLOG_COMPRESSION": "zstd"}), nil); err == nil {
		t.Error("zstd accepted")
	}
	_, warns, err := loadConfig(env(map[string]string{"OPENLOG_SAMPLING_RATIO": "abc", "OPENLOG_METRIC_EXPORT_INTERVAL": "-1s"}), nil)
	if err != nil {
		t.Fatal(err)
	}
	joined := strings.Join(warns, "\n")
	for _, want := range []string{"OPENLOG_SAMPLING_RATIO", "OPENLOG_METRIC_EXPORT_INTERVAL", "license key", "service name"} {
		if !strings.Contains(joined, want) {
			t.Errorf("warnings %q lack %q", joined, want)
		}
	}
}

func eq[T comparable](t *testing.T, got, want T) {
	t.Helper()
	if got != want {
		t.Errorf("got %v, want %v", got, want)
	}
}

File: agents/node/test/unit/config.test.ts
import assert from 'node:assert/strict';
import { test } from 'node:test';
import { ConfigError, LICENSE_KEY_HEADER, loadConfig, parseGoDuration, parseKV } from '../../src/config';

test('defaults', () => {
  const { config: c, warnings } = loadConfig({}, {});
  assert.equal(c.enabled, true);
  assert.equal(c.endpoint, 'http://localhost:4318');
  assert.equal(c.protocol, 'http/protobuf');
  assert.equal(c.compression, 'gzip');
  assert.equal(c.samplingRatio, 1);
  assert.equal(c.samplingRV, false);
  assert.equal(c.logLevel, 'warn');
  assert.equal(c.metricIntervalMs, 60_000);
  assert.equal(c.shutdownTimeoutMs, 5_000);
  assert.equal(c.dbQueryText, 'sanitized');
  assert.equal(c.infraRuntimeDir, '/run/openlog-infra-agent');
  assert.equal(c.infraStateDir, '/var/lib/openlog-infra-agent');
  assert.match(c.serviceName, /^unknown_service:/);
  assert.ok(warnings.some((w) => w.includes('OPENLOG_SERVICE_NAME')));
  assert.ok(warnings.some((w) => w.includes('OPENLOG_LICENSE_KEY')));
});

test('precedence: options > OPENLOG_* > OTEL_*', () => {
  const env = {
    OTEL_EXPORTER_OTLP_ENDPOINT: 'http://otel:4318',
    OTEL_SERVICE_NAME: 'otel-name',
    OTEL_EXPORTER_OTLP_HEADERS: 'x-a=1,x-b=hello%20world',
    OTEL_RESOURCE_ATTRIBUTES: 'team=a,service.namespace=ns',
    OTEL_TRACES_SAMPLER: 'parentbased_traceidratio',
    OTEL_TRACES_SAMPLER_ARG: '0.5',
    OTEL_METRIC_EXPORT_INTERVAL: '15000',
    OPENLOG_ENDPOINT: 'https://ingest.example.com/',
    OPENLOG_SERVICE_NAME: 'checkout',
    OPENLOG_LICENSE_KEY: 'k1',
    OPENLOG_SAMPLING_RATIO: '0.25',
    OPENLOG_SAMPLING_RV: 'true',
    OPENLOG_RESOURCE_ATTRIBUTES: 'team=b',
    OPENLOG_METRIC_EXPORT_INTERVAL: '1m30s',
    OPENLOG_SHUTDOWN_TIMEOUT: '750ms',
    OPENLOG_HOST_ID: 'host-12345678',
    OPENLOG_DB_QUERY_TEXT: 'raw',
    OPENLOG_INSTRUMENTATIONS_DISABLED: 'aws-sdk, graphql',
  };
  const { config: c } = loadConfig(env, { serviceVersion: '1.2.3', headers: { [LICENSE_KEY_HEADER]: 'ignored' } });
  assert.equal(c.endpoint, 'https://ingest.example.com');
  assert.equal(c.serviceName, 'checkout');
  assert.equal(c.serviceVersion, '1.2.3');
  assert.equal(c.headers['x-a'], '1');
  assert.equal(c.headers['x-b'], 'hello world');
  assert.equal(c.headers[LICENSE_KEY_HEADER], 'k1'); // license key wins over headers
  assert.equal(c.resourceAttributes.team, 'b');
  assert.equal(c.resourceAttributes['service.namespace'], 'ns');
  assert.equal(c.samplingRatio, 0.25);
  assert.equal(c.samplingRV, true);
  assert.equal(c.metricIntervalMs, 90_000);
  assert.equal(c.shutdownTimeoutMs, 750);
  assert.equal(c.hostId, 'host-12345678');
  assert.equal(c.dbQueryText, 'raw');
  assert.deepEqual(c.disabledInstrumentations, ['aws-sdk', 'graphql']);

  const o = loadConfig(env, { samplingRatio: 0.1, endpoint: 'http://x:1' }).config;
  assert.equal(o.samplingRatio, 0.1);
  assert.equal(o.endpoint, 'http://x:1');
  assert.equal(loadConfig({ OTEL_TRACES_SAMPLER_ARG: '0.5' }).config.samplingRatio, 0.5);
  assert.equal(loadConfig({ OTEL_TRACES_SAMPLER: 'always_on', OTEL_TRACES_SAMPLER_ARG: '0.5' }).config.samplingRatio, 1);
  assert.equal(loadConfig({ OTEL_METRIC_EXPORT_INTERVAL: '15000' }).config.metricIntervalMs, 15_000);
});

test('enabled flags', () => {
  assert.equal(loadConfig({ OTEL_SDK_DISABLED: 'true' }).config.enabled, false);
  assert.equal(loadConfig({ OTEL_SDK_DISABLED: 'true', OPENLOG_ENABLED: 'true' }).config.enabled, true);
  assert.equal(loadConfig({ OPENLOG_ENABLED: '0' }).config.enabled, false);
  const { config, warnings } = loadConfig({ OPENLOG_ENABLED: 'nope' });
  assert.equal(config.enabled, true);
  assert.ok(warnings.some((w) => w.includes('OPENLOG_ENABLED')));
});

test('validation and normalization', () => {
  assert.throws(() => loadConfig({ OPENLOG_PROTOCOL: 'http/json' }), ConfigError);
  assert.throws(() => loadConfig({ OPENLOG_COMPRESSION: 'zstd' }), ConfigError);
  assert.throws(() => loadConfig({ OPENLOG_ENDPOINT: 'ftp://x' }), ConfigError);
  assert.equal(loadConfig({ OPENLOG_PROTOCOL: 'grpc' }).config.endpoint, 'http://localhost:4317');
  assert.equal(loadConfig({ OPENLOG_ENDPOINT: 'ingest.example.com:4318' }).config.endpoint, 'https://ingest.example.com:4318');
  assert.equal(loadConfig({ OPENLOG_COMPRESSION: 'none' }).config.compression, 'none');
  const clamped = loadConfig({ OPENLOG_SAMPLING_RATIO: '1.5' });
  assert.equal(clamped.config.samplingRatio, 1);
  assert.ok(clamped.warnings.some((w) => w.includes('clamped')));
  const bad = loadConfig({ OPENLOG_SAMPLING_RATIO: 'abc', OPENLOG_METRIC_EXPORT_INTERVAL: '10', OPENLOG_LOG_LEVEL: 'loud' });
  assert.equal(bad.config.samplingRatio, 1);
  assert.equal(bad.config.metricIntervalMs, 60_000);
  assert.equal(bad.warnings.length >= 3, true);
  assert.equal(loadConfig({ OTEL_RESOURCE_ATTRIBUTES: 'service.name=from-attrs' }).config.serviceName, 'from-attrs');
  // Authorization header instead of a license key: no warning
  assert.ok(!loadConfig({ OTEL_EXPORTER_OTLP_HEADERS: 'Authorization=Bearer x', OPENLOG_SERVICE_NAME: 's' }).warnings.length);
});

test('parseKV and parseGoDuration', () => {
  assert.deepEqual(parseKV(' a = 1 ,b=%3D%2C, =x,novalue,c=%zz'), { a: '1', b: '=,', c: '%zz' });
  assert.equal(parseGoDuration('60s'), 60_000);
  assert.equal(parseGoDuration('1h2m3.5s'), 3_723_500);
  assert.equal(parseGoDuration('250ms'), 250);
  assert.equal(parseGoDuration('1500us'), 1.5);
  assert.equal(parseGoDuration('0'), 0);
  assert.equal(parseGoDuration('10'), undefined);
  assert.equal(parseGoDuration('5 s'), undefined);
  assert.equal(parseGoDuration(''), undefined);
});
